File: tests/typesafety/typescript/exercise_user_json_start.ts
// expect: passes

function greeting(user: Record<string, unknown>): string {
  const name = user["name"];
  if (typeof name !== "string") {
    throw new Error("name missing");
  }
  return `hello ${name}`;
}

function canVote(user: Record<string, unknown>): boolean {
  const age = user["age"];
  if (typeof age !== "number") {
    throw new Error("age missing");
  }
  return age >= 18;
}

export function summary(body: string): string {
  const user: unknown = JSON.parse(body);
  if (typeof user !== "object" || user === null) {
    throw new Error("not an object");
  }
  return `${greeting(user as Record<string, unknown>)}, can vote: ${canVote(user as Record<string, unknown>)}`;
}

File: tests/typesafety/python/exercise_user_json_start.py
# expect: passes
# title: greeting, can_vote and summary all re-check the user dict
import json


def greeting(user: dict[str, object]) -> str:
    name = user.get("name")
    if not isinstance(name, str):
        raise ValueError("name missing")
    return f"hello {name}"


def can_vote(user: dict[str, object]) -> bool:
    age = user.get("age")
    if not isinstance(age, int):
        raise ValueError("age missing")
    return age >= 18


def summary(body: str) -> str:
    user = json.loads(body)
    if not isinstance(user, dict):
        raise ValueError("not an object")
    return f"{greeting(user)}, can vote: {can_vote(user)}"
